### ISSUE TITLE
Sequel performs bad on rails when used as primary db without preload_app! parameter and w/ threads

### ISSUE THREAD

**Issue by gencer** [jeremyevans/sequel#1354]
As I started a issue on https://github.com/TalentBox/sequel-rails/issues/137, I finally find myself by using initializer for my db and extend models to a abstract class.

Due to I do not use preload_app! on **puma** in needed of phased restart (Zero downtime) ability, I use without it. I do have prune_bundler enabled and preload disabled. Due to this, events on worker boot and shutdown does not work for me.

If I make Sequel my primary ORM and remove ActiveRecord, And If I add 100-200 pool on each worker, It just goes insane. Usually I get Cannot allocate memory errors across my device. Need to restart. ActiveRecord, however, does not have this issue. Everytime i start a benchmark one by one, it just fast and without compromise. Either there is something wrong with Sequel or Rails. Sequel either opens too much connection or does not see my settings properly.

Hovewer, hence this issue persist, I created an initializer inside of a rails. I put manually this code:

```
DB = YAML::load(ERB.new(File.read(Rails.root.join("config","database_mail.yml"))).result)[Rails.env]
Sequel::Model.db.extension(:connection_validator)
Sequel::Model.db.pool.connection_validation_timeout = -1
```

With this code I also created a module (abstract class for my models):

```
module MasterModel
  def self.Model(source)
    Rails.logger.debug 'Worker connected to DB'
    c = Sequel::Model(
        Sequel.connect(DB)
        )
    c.set_dataset(source)
  end
end
```
And of course instead of `Sequel::Model`, I extend my models to `MasterModel::Model(:table_name`).

Now, I executed this code on each request:

`MasterModel::Mail.select(:some_field).where(:id => 5).first` instead of direct model.

Before this way, When I do `wrk` test 3 times one by one, I get 150 reqs/s in first, 60 reqs/s in second and 0req/s on third and after third my system goes down. But after this above implementation, I just started to get 3500 reqs/s. Not on first but almost 20th of test. Everytime I run it goes more.

Fact is, Same code with primary db is not work like this. I found some docs that says threaded apps should have initializers but also same doc indicates that if we have Rails 4.1+ we should leave it to database.yml. But it just doesn't work as expected.

If you see above code you will notice `Rails.logger`. I looked to log file and see that only 24 lines of code has been added. So Sequel successfully established connections for 24 workers and not more. They are just fine.

Does Sequel something that I do not know on this issue? What should I do to fix this? Is this a **bug?** Or **feature?** (expected behavior)

Only 4 things came to my mind on this:

1. Sequel does not start or establish an connection and pool them after workers start if we do not enable `preload_app`
2. Sequel connects dozens of times instead of each worker.
3. I do something wrong.
4. In my way, there is no pools created. Same connection used everytime. _But this does not explain how my threaded apps handle such too much concurrency with db select._

**Comment by gencer** [comment]
I see that, when `max_connections `parameter provided with **SequelRails**, My app goes down by requests per second. If I remove max_connection it just works as primary.

Is it wise to remove this parameter for multi-processed multi-threaded apps?

**Comment by gencer** [comment]
Yep, this was caused high number of connections per thread.

**Comment by jeremyevans** [comment]
This was probably caused because you are creating separate `Sequel::Database` objects in each call to `MasterModel.Model`.  Change your initializer to do:

```ruby
DB = Sequel.connect(YAML::load(ERB.new(File.read(Rails.root.join("config","database_mail.yml"))).result)[Rails.env])
DB.extension(:connection_validator)
DB.pool.connection_validation_timeout = -1
```

And your MasterModel.Model code to:

```ruby
module MasterModel
  def self.Model(source)
    Rails.logger.debug 'Worker connected to DB'
    c = Sequel::Model(DB)
    c.set_dataset(source)
  end
end
```

Then you'll have a single `Sequel::Database` object, and everything should work fine, even with a high number of max_connections.

**Comment by gencer** [comment]
Oh, this is awesome. Thanks. Let me test this now.

BTW, If we only init once and use with all, why does my primary connection (by using sequel + sequel_rails gem) does not work on high number of max_connections?

I can live with this MasterModel way but would like to use standardized way.

**Comment by gencer** [comment]
@jeremyevans, I tested your solution. It does not work with high number of max_connection.

```
Running 10s test @ **
  32 threads and 150 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    29.90ms   31.11ms 133.26ms   82.87%
    Req/Sec    50.84     35.31   151.00     71.43%
  286 requests in 10.10s, 153.33KB read
Requests/sec:     28.32       <-- This is what I get with max_connections = 100. If I give 4 this becames 3000~
Transfer/sec:     15.18KB
```

Do you have any idea? Otherwise I will do;

```
On puma:
24 workers
200 Threads
No preload
4 max_connections. (was 200, same number as thread count)
```

This config just works without initializer.

**Comment by gencer** [comment]
@jeremyevans one more question. Could be this slowness be happen because of pgpool2 extension? Currently, I connect to pgpool and then to postgresql.

I couldn't find any documentation like "do not use another pooling app with sequel" warning. So could be this a problem?

Should I directly connected to postgresql?

**Comment by jeremyevans** [comment]
24 workers * 100 connections is 2400 connections, which is probably taxing your database server way too much.  I would stay at 4 connections/worker, and only increase if you have benchmarked and found a higher number to be faster.  Instead of testing at 100, try testing at 6 or 8 and seeing what the differences are compared to 4.

Due to the GVL, unless your workload is purely IO bound, 200 threads is probably at least an order of magnitude too high.  I would keep your puma at like 2-4 threads/worker, and again, only increase if you have benchmarked and found it faster.

**Comment by gencer** [comment]
@jeremyevans thanks for the explanation. Yes, you're right. I also made a very simple mistake. I did 2 different app to connect same database and do pooling on it. Due to this it just goes down. I seperated apps, lowered threads and used even high number of max_connections (higher than 4) for each one, dropped pgpool and directly connected to postgresql's 5432 port. It just works. I do stress tests. Even multiple tests at the same time. No slowdown, nothing.

It's all about misconfiguration and also it was a bad idea to use sequel's own pool system with pgpool. This also adds overhead to response time. After switched to 5432 port, it becomes more responsive.

Thanks,
Gencer.
